vortex-serde/src/layouts/pruning.rs

```diff
@@ -0,0 +1,538 @@
+use std::collections::hash_map::Entry;
+use std::sync::Arc;
+
+use ahash::{HashMap, HashMapExt};
+use vortex::stats::Stat;
+use vortex_dtype::field::Field;
+use vortex_dtype::Nullability;
+use vortex_expr::{BinaryExpr, Column, Literal, Operator, VortexExpr};
+use vortex_scalar::Scalar;
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct ColumnStat {
+    column: Field,
+    stat: Stat,
+}
+
+impl ColumnStat {
+    pub fn new(column: Field, stat: Stat) -> Self {
+        Self { column, stat }
+    }
+
+    pub fn column(&self) -> &Field {
+        &self.column
+    }
+
+    pub fn stat(&self) -> Stat {
+        self.stat
+    }
+}
+
+#[allow(dead_code)]
+pub struct PruningPredicate {
+    expr: Arc<dyn VortexExpr>,
+    stat_references: HashMap<Field, ColumnStat>,
+}
+
+impl PruningPredicate {
+    pub fn new(original_expr: &Arc<dyn VortexExpr>) -> Self {
+        let (expr, stat_references) = convert_to_pruning_expression(original_expr);
+        Self {
+            expr,
+            stat_references,
+        }
+    }
+
+    pub fn references(&self) -> &HashMap<Field, ColumnStat> {
+        &self.stat_references
+    }
+
+    pub fn necessary_statistics(&self) -> HashMap<Field, Vec<Stat>> {
+        let mut to_fetch: HashMap<Field, Vec<Stat>> = HashMap::new();
+        for column_stat in self.stat_references.values() {
+            match to_fetch.entry(column_stat.column().clone()) {
+                Entry::Occupied(o) => o.into_mut().push(column_stat.stat()),
+                Entry::Vacant(v) => {
+                    v.insert(vec![column_stat.stat()]);
+                }
+            }
+        }
+        to_fetch
+    }
+}
+
+fn convert_to_pruning_expression(
+    expr: &Arc<dyn VortexExpr>,
+) -> (Arc<dyn VortexExpr>, HashMap<Field, ColumnStat>) {
+    // Anything that can't be translated has to be represented as
+    // boolean true expression, i.e. the value might be in that chunk
+    let fallback = Arc::new(Literal::new(Scalar::bool(true, Nullability::NonNullable)));
+    // TODO(robert): Add support for boolean column expressions,
+    //  i.e. if column is of bool dtype it's valid to filter on it directly as a predicate
+    if expr.as_any().downcast_ref::<Column>().is_some() {
+        return (fallback, HashMap::new());
+    }
+
+    if let Some(bexp) = expr.as_any().downcast_ref::<BinaryExpr>() {
+        if bexp.op() == Operator::Or || bexp.op() == Operator::And {
```

Comment:
maybe `op.is_logical()`?

Reply:
Not sure, `!` (Not) would be logical but shouldn't be handled here. This is specifically just And and Or

Reply:
fair enough